src/main/scala/is/hail/methods/PCA.scala

```diff
@@ -53,15 +54,63 @@ trait PCA {
       new KeyTable(vsm.hc, rdd, rowType, Array("v"))
     })
 
-    (svd.V.multiply(DenseMatrix.diag(svd.s)), optionLoadings, someIf(computeEigenvalues, svd.s.toArray.map(math.pow(_, 2))))
+    (svd.s.toArray.map(math.pow(_, 2)), svd.V.multiply(DenseMatrix.diag(svd.s)), optionLoadings)
   }
 
-  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Variant]], IndexedRowMatrix)
-}
+  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, expr: String, getVariants: Boolean): (Option[Array[Any]], IndexedRowMatrix) = {
+
+    val partitionSizes = vsm.rdd2.mapPartitions(it => Iterator.single(it.size)).collect().scanLeft(0)(_ + _)
+    assert(partitionSizes.length == vsm.rdd2.getNumPartitions + 1)
+    val pSizeBc = vsm.sparkContext.broadcast(partitionSizes)
+
+    val rowType = vsm.rowType
+    val samplesBc = vsm.sampleIdsBc
+    val sampleAnnotationsBc = vsm.sampleAnnotationsBc
+
+    val ec = EvalContext(Map(
+      "global" -> (0, vsm.globalSignature),
+      "v" -> (1, vsm.vSignature),
+      "va" -> (2, vsm.vaSignature),
+      "s" -> (3, vsm.sSignature),
+      "sa" -> (4, vsm.saSignature),
+      "g" -> (5, vsm.genotypeSignature)))
+    val f = RegressionUtils.parseExprAsDouble(expr, ec)
+    ec.set(0, vsm.globalAnnotation)
 
-object SamplePCA extends PCA {
-  override def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Variant]], IndexedRowMatrix) = {
-    val (variants, mat) = ToHWENormalizedIndexedRowMatrix(vsm)
-    (if (getVariants) Option(variants) else None, mat)
+    val mat = vsm.rdd2.mapPartitionsWithIndex {case (i, it) =>
+      val pStartIdx = pSizeBc.value(i)
+      var j = 0
+      val ur = new UnsafeRow(rowType)
+      it.map {rv =>
+        ur.set(rv)
+        ec.set(1, ur.get(1))
+        ec.set(2, ur.get(2))
+        val gs = ur.getAs[IndexedSeq[Any]](3)
+        val ns = gs.length
+        val a = new Array[Double](ns)
+        var k = 0
+        while(k < ns) {
+          ec.set(3, samplesBc.value(k))
+          ec.set(4, sampleAnnotationsBc.value(k))
+          ec.set(5, gs(k))
+          a(k) = f().toDouble
```

Comment:
You should check `f()` is not `null` here.  I assume `expr` must be defined, so you should throw an error.  (Once the expr language propagates missingness, we can just require the expr is non-missing.)